provider/clusterrosa/hcp/upgrade/upgrade.go

```diff
@@ -0,0 +1,236 @@
+	for _, v := range version.AvailableUpgrades() {
+		id := ocmUtils.CreateVersionId(v, version.ChannelGroup())
+		resp, err := versionClient.Version(id).
+			Get().
+			Send()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get version information: %v", err)
+		}
+		availableVersion := resp.Body()
+		if availableVersion.HostedControlPlaneEnabled() {
+			availableUpgradeVersions = append(availableUpgradeVersions, availableVersion)
+		}
+	}
+
+	return availableUpgradeVersions, nil
+}
+
+// Get the list of upgrade policies associated with a cluster
+func GetScheduledUpgrades(ctx context.Context, client *cmv1.ClustersClient, clusterId string) ([]ControlPlaneUpgrade, error) {
+	upgrades := []ControlPlaneUpgrade{}
+
+	// Get the upgrade policies for the cluster
+	upgradePolicies := []*cmv1.ControlPlaneUpgradePolicy{}
+	upgradeClient := client.Cluster(clusterId).ControlPlane().UpgradePolicies()
+	page := 1
+	size := 100
+	for {
+		resp, err := upgradeClient.List().
+			Page(page).
+			Size(size).
+			SendContext(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to list control plane upgrade policies: %v", err)
+		}
+		upgradePolicies = append(upgradePolicies, resp.Items().Slice()...)
+		if resp.Size() < size {
+			break
+		}
+		page++
+	}
+
+	// For each upgrade policy, get its state
+	for _, policy := range upgradePolicies {
+		if policy.UpgradeType() != cmv1.UpgradeTypeControlPlane {
+			continue
+		}
+		resp, err := upgradeClient.ControlPlaneUpgradePolicy(policy.ID()).
+			Get().
+			SendContext(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get upgrade policy state: %v", err)
+		}
+		state := resp.Body().State()
+		upgrades = append(upgrades, ControlPlaneUpgrade{
+			Policy:      policy,
+			PolicyState: state,
+		})
+	}
+
+	return upgrades, nil
+}
+
+// Check the provided list of upgrades, canceling pending upgrades that are not
+// for the correct version, and returning an error if there is already an
+// upgrade in progress that is not for the desired version
+func CheckAndCancelUpgrades(
+	ctx context.Context, client *cmv1.ClustersClient, upgrades []ControlPlaneUpgrade, desiredVersion *semver.Version) (bool, error) {
+	correctUpgradePending := false
+	tenMinFromNow := time.Now().UTC().Add(10 * time.Minute)
+
+	for _, upgrade := range upgrades {
+		tflog.Debug(ctx, fmt.Sprintf("Found existing upgrade policy to '%s' in state '%s'", upgrade.Policy.Version(), upgrade.PolicyState.Value()))
+		toVersion, err := semver.NewVersion(upgrade.Policy.Version())
+		if err != nil {
+			return false, fmt.Errorf("failed to parse upgrade version: %v", err)
+		}
+		switch upgrade.PolicyState.Value() {
+		case cmv1.UpgradePolicyStateValueDelayed, cmv1.UpgradePolicyStateValueStarted:
+			tflog.Debug(ctx, fmt.Sprintf("---- %v", desiredVersion.Equal(toVersion)))
+			if desiredVersion.Equal(toVersion) {
+				correctUpgradePending = true
+			} else {
+				return false, fmt.Errorf("a cluster upgrade is already in progress")
+			}
```

Comment:
Early exit is more readable
```suggestion
			if !desiredVersion.Equal(toVersion) {
				return false, fmt.Errorf("a cluster upgrade is already in progress")
			} 	
			correctUpgradePending = true
```